### PR TITLE
Add native AOT runtime-deps and sdk variants

### PR DESCRIPTION
**Pull request by lbussell**
Please see https://github.com/dotnet/dotnet-docker/issues/4129 and https://github.com/dotnet/dotnet-docker/issues/4804

~~I am currently working on the tests for these images.~~ After discussion, tests will be included in a separate PR.

### CODE REVIEW

File: src/runtime-deps/8.0/cbl-mariner2.0-distroless/arm64v8/Dockerfile [hunk 1]
```diff
@@ -13,7 +13,6 @@ RUN mkdir /staging \
         \
         # .NET dependencies
         glibc \
-        krb5 \
```
Comment (MichaelSimons) [inline src/runtime-deps/8.0/cbl-mariner2.0-distroless/arm64v8/Dockerfile:16]: Is the reason this is being removed captured in an issue?

File: eng/dockerfile-templates/Dockerfile.linux.install-deps [hunk 2]
```diff
@@ -1,30 +1,41 @@
 {{
     _ ARGS:
-        isSdk (optional): Indicates whether the dependencies are installed for an sdk Dockerfile
+        is-aot (optional): TODO: fill out
```
Comment (MichaelSimons) [inline eng/dockerfile-templates/Dockerfile.linux.install-deps:3]: What are the plans to address the TODO comments here?

File: eng/dockerfile-templates/runtime-deps/Dockerfile.chiseled-ubuntu [hunk 3]
```diff
@@ -11,6 +16,31 @@
...
+            "libgcc-s1_libs",
+            "libssl3_libs",
+            "zlib1g_libs"
+        ] ^
+
+    _ Arrays in cottle are actually key-value maps. Sorting the array sorts by value
+        but does not change the key value. In order to have the right index in the
+        array, we must call `cat` again to reset the indices (keys) in the map. ^
+    set standardPkgs to cat(sort(cat(aotPkgs, [ "libstdc++6_libs" ]))) ^
+    set extraPkgs to cat(sort(cat(standardPkgs,
```
Comment (MichaelSimons) [inline eng/dockerfile-templates/runtime-deps/Dockerfile.chiseled-ubuntu:33]: I see support for extraPkgs but no invocation.  I understood this was going to be handled with a separate PR.  Is this a case of separating the two becomes difficult because they are touching the same template code?

File: src/sdk/8.0/jammy-aot/amd64/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+deb [arch=arm64] http://ports.ubuntu.com/ubuntu-ports/ jammy-updates main restricted
+EOF
+
+RUN sed -i -e 's/deb http/deb [arch=amd64] http/g' /etc/apt/sources.list \
+    && dpkg --add-architecture arm64 \
+    && apt-get update \
+    && apt-get install -y --no-install-recommends \
+        clang \
+        gcc-aarch64-linux-gnu \
+        llvm \
```
Comment (MichaelSimons) [inline src/sdk/8.0/jammy-aot/amd64/Dockerfile:16]: Why LLVM?

File: src/runtime-deps/6.0/jammy-chiseled/amd64/Dockerfile [hunk 5]
```diff
@@ -34,7 +34,7 @@ RUN groupadd \
 
 RUN mkdir -p /rootfs/var/lib/dpkg/ \
     && chisel-wrapper --generate-dpkg-status /rootfs/var/lib/dpkg/status -- \
-        --release "ubuntu-22.04" --root /rootfs \
+        --release ubuntu-22.04 --root /rootfs \
```
Comment (mthalman) [inline src/runtime-deps/6.0/jammy-chiseled/amd64/Dockerfile:37]: Because this is making a change that would also be relevant to the main branch, this change should be made independently of this PR. It will get in the way of merging changes between the branches and dealing with diffs.

File: src/runtime-deps/8.0/cbl-mariner2.0-distroless/arm64v8/Dockerfile (hunk 1 above)
Comment (mthalman) [reply src/runtime-deps/8.0/cbl-mariner2.0-distroless/arm64v8/Dockerfile:16]: https://github.com/dotnet/dotnet-docker/issues/4751. IMO, this removal should be done separately from these changes. The same removal should apply to Debian and Ubuntu, right? https://github.com/dotnet/dotnet-docker/blob/a345a8f9085bd933c2850cbb4e1d203f5a23a987/src/runtime-deps/8.0/bookworm-slim/amd64/Dockerfile#L18

I understand it's installed by default there. We should have discussion of whether it should be listed in that case.

**Review by mthalman — COMMENTED** [review]
My meta-review comment is "let's be sure to avoid including changes in this PR that we want in main" because AOT won't be getting merged to main.

File: src/sdk/8.0/jammy-aot/amd64/Dockerfile (hunk 4 above)
Comment (lbussell) [reply src/sdk/8.0/jammy-aot/amd64/Dockerfile:16]: I referenced the package list from Rich's sample - https://github.com/dotnet/dotnet-docker/blob/58ee815e1a06614732a2c2726bec51a0c9b8ad67/samples/releasesapi-aot/Dockerfile.ubuntu-cross-x64-arm64#L14

@richlander is LLVM strictly necessary for cross-compilation scenarios?

File: src/runtime-deps/8.0/cbl-mariner2.0-distroless/arm64v8/Dockerfile (hunk 1 above)
Comment (lbussell) [reply src/runtime-deps/8.0/cbl-mariner2.0-distroless/arm64v8/Dockerfile]: I have removed this change and will include it in the `extra` PR when we'll have the full set of images available.

File: src/sdk/8.0/jammy-aot/amd64/Dockerfile (hunk 4 above)
Comment (MichaelSimons) [reply src/sdk/8.0/jammy-aot/amd64/Dockerfile:16]: This feels related to a comment Matt made at https://github.com/dotnet/dotnet-docker/pull/4822#discussion_r1302102256.  If LLVM is required, should we be explicitly declaring it as a dependency in all distros even in scenarios where it is already provided in a base layer?

File: src/runtime-deps/8.0/cbl-mariner2.0-distroless/arm64v8/Dockerfile (hunk 1 above)
Comment (richlander) [reply src/runtime-deps/8.0/cbl-mariner2.0-distroless/arm64v8/Dockerfile]: We should have the discussion on whether we want to include this in `extra`. I don't think so. My understanding is that kerberos is often paired with ldap, which we have never installed.

File: src/sdk/8.0/jammy-aot/amd64/Dockerfile (hunk 4 above)
Comment (richlander) [reply src/sdk/8.0/jammy-aot/amd64/Dockerfile:16]: `clang` depends on `llvm` so it will be pulled in anyway.

Clang is the front-end and LLVM is the backend.

Comment (mthalman) [reply src/sdk/8.0/jammy-aot/amd64/Dockerfile:16]: In that case, I lean towards only listing `clang` for now. Depending on CVE activity in dependent packages, we may want to include others but I think we can evaluate that over time.

File: src/sdk/8.0/jammy-aot/arm64v8/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# syntax=docker/dockerfile:1
+ARG REPO=mcr.microsoft.com/dotnet/sdk
+FROM $REPO:8.0.100-rc.1-jammy-arm64v8
+
+COPY <<EOF /etc/apt/sources.list.d/amd64.list
```
Comment (MichaelSimons) [inline src/sdk/8.0/jammy-aot/arm64v8/Dockerfile:5]: Do we have a general agreement that HereDocs should be used?  If so what is the thought on when it should be used?  I would expect consistent usage of it - e.g. in RUN instructions as well.  Making consistent use is beyond the scope of this PR but given the introduction of it here we should establish an agreed upon direction and log follow-ups tracking this work.

Comment (lbussell) [reply src/sdk/8.0/jammy-aot/arm64v8/Dockerfile:5]: Thanks for bringing this up. I tried heredocs for RUN instructions but they don't play nice with line endings with Git for Windows' default behavior (checkout as CRLF and commit as LF). It creates some really strange errors.

Because of that I'd probably lean towards switching this command to using `echo` instead, to keep everything consistent.

Comment (MichaelSimons) [reply src/sdk/8.0/jammy-aot/arm64v8/Dockerfile:5]: I agree that sounds best for now.

File: src/sdk/8.0/jammy-aot/arm64v8/Dockerfile [hunk 7]
```diff
@@ -2,10 +2,8 @@
 ARG REPO=mcr.microsoft.com/dotnet/sdk
 FROM $REPO:8.0.100-rc.1-jammy-arm64v8
 
-COPY <<EOF /etc/apt/sources.list.d/amd64.list
-deb [arch=amd64] http://archive.ubuntu.com/ubuntu/ jammy main restricted
-deb [arch=amd64] http://archive.ubuntu.com/ubuntu/ jammy-updates main restricted
-EOF
+RUN echo "deb [arch=amd64] http://archive.ubuntu.com/ubuntu/ jammy main restricted" > /etc/apt/sources.list.d/amd64.list \
```
Comment (MichaelSimons) [inline src/sdk/8.0/jammy-aot/arm64v8/Dockerfile:5]: Now that this utilizes echo, it doesn't feel necessary to keep this as a separate instruction especially given the first line of the next RUN instruction.

**Review by mthalman — APPROVED** [review]

**Review by MichaelSimons — APPROVED** [review]

File: src/sdk/8.0/jammy-aot/arm64v8/Dockerfile (hunk 6 above)
Comment (richlander) [reply src/sdk/8.0/jammy-aot/arm64v8/Dockerfile]: I noticed the same thing. You have to do unix-style to use heredocs.

Line endings are the most annoying (at least obvious) difference between Windows and unix-style. More so than slash/backslash.

**Comment by lbussell** [comment]
This PR and https://github.com/dotnet/dotnet-docker/pull/4847 have significant overlap in the templates. I'd like to wait until that is merged first before merging this one so that the merge of that PR to main is less complicated.

**Review by mthalman — APPROVED** [review]
